syntax = "proto3";

package raft;

service RaftService {
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  rpc GetStatus(GetStatusRequest) returns (GetStatusResponse);
}

message RequestVoteRequest {
  int32 term = 1;
  string candidate_id = 2;
  int32 last_log_index = 3;
  int32 last_log_term = 4;
}

message RequestVoteResponse {
  int32 term = 1;
  bool vote_granted = 2;
  string from_node = 3;
}

message LogEntry {
  int32 term = 1;
  int32 index = 2;
  string command = 3;
}

message AppendEntriesRequest {
  int32 term = 1;
  string leader_id = 2;
  int32 prev_log_index = 3;
  int32 prev_log_term = 4;
  repeated LogEntry entries = 5;
  int32 leader_commit = 6;
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
  string from_node = 3;
}

message GetStatusRequest {}

message GetStatusResponse {
  string node_id = 1;
  string state = 2;
  int32 term = 3;
  string leader_id = 4;
  int32 log_size = 5;
  int32 commit_index = 6;
  int32 last_applied = 7;
  int32 cluster_size = 8;
}
